flashlight/distributed/reducers/CoalescingReducer.cpp

```diff
@@ -22,6 +22,7 @@ CoalescingReducer::~CoalescingReducer() {
 }
 
 void CoalescingReducer::add(Variable& var) {
+  var.cast(f32);
```

Comment:
Precisely. However, there is a bit complication involved. That's why I initially moved ahead with a naive implementation: 
1. In a lazy approach, a cast to a higher precision will increase the current size of `cache_` by ~2X. The new size can be larger than the the constant threshold that is set for flushing the data (i.e., `kCoalesceCacheSize`). This is gonna break other assumptions regarding when we actually flush, since flushing can happen anywhere in the range of `[kCoalesceCacheSize, 2*kCoalesceCacheSize)`.
2. We can find a remedy for the aforementioned issue, but also we need to reconsider the choices that are made for coalesced memory allocation in `NcclContext::getCoalesceBuffer()`.

Perhaps we can keep this in the naive form for this PR in favor of agility.

Reply:
These are great points. I do have ways to resolve both of these issues, but you're right that for now, we can move forward with a naive implementation (Y) and make those other changes down the line.